File: EasyLibrary.Application/Services/BookTypesService.cs
using EasyLibrary.Application.Exceptions;
using EasyLibrary.Core;
using EasyLibrary.Core.Abstractions;
using EasyLibrary.Core.Models;
using EasyLibrary.DTO.BookTypes;

namespace EasyLibrary.Application.Services
{
    public class BookTypesService : IBookTypesService
    {
        private readonly IBookAuthorsRepository _bookAuthorsRepository;
        private readonly IBookSeriesRepository _bookSeriesRepository;
        private readonly IBookTypesRepository _bookTypesRepository;
        private readonly IPublishingHouseRepository _publishingHouseRepository;

        public BookTypesService(
            IBookAuthorsRepository bookAuthorsRepository,
            IBookSeriesRepository bookSeriesRepository, 
            IBookTypesRepository bookTypesRepository, 
            IPublishingHouseRepository publishingHouseRepository
           )
        {
            _bookAuthorsRepository = bookAuthorsRepository;
            _bookSeriesRepository = bookSeriesRepository;
            _bookTypesRepository = bookTypesRepository;
            _publishingHouseRepository = publishingHouseRepository;
        }

        public async Task<Guid> CreateBookType(BookTypeRequest request)
        {
            var bookType = await MapToBookType(request);

            return await _bookTypesRepository.Create(bookType); 
        }

        public async Task<Guid> DeleteBookType(Guid id)
        {
            return await _bookTypesRepository.Delete(id);
        }

        public async Task<List<BookType>> GetAllBookTypes()
        {
            return await _bookTypesRepository.Get();
        }

        public async Task<BookType> GetBookTypeById(Guid id)
        {
            var bookType = await _bookTypesRepository.GetById(id);

            return bookType ?? throw new NotFoundException<BookType>(id);
        }

        public async Task<Guid> UpdateBookType(Guid id, BookTypeRequest request)
        {
            var bookType = await MapToBookType(request, id);
            return await _bookTypesRepository.Update(bookType); 
        }

        private async Task<BookType> MapToBookType(BookTypeRequest request, Guid id = default)
        {
            var publishingHouse =
                await _publishingHouseRepository.GetById(request.PublishingHouseId) ??
                throw new NotFoundException<PublishingHouse>(request.PublishingHouseId);

            var bookSeries =
                await _bookSeriesRepository.GetById(request.BookSeriesId) ??
                throw new NotFoundException<BookSeries>(request.BookSeriesId);

            var authors = new List<BookAuthor>(); 

            foreach(var authorId in request.AuthorsId)
            {
                authors.Add(await _bookAuthorsRepository.GetById(authorId) ??
                    throw new NotFoundException<BookAuthor>(authorId));    
            }

            var bookType = BookType.Create(
                id = id == default? Guid.NewGuid() : id,
                request.Title,
                publishingHouse,
                bookSeries,
                new List<BookCopy>(),
                authors.ToList(),
                (BookBinding)request.Cover,
                request.PublishingYear,
                request.ISBN,
                request.PagesCount,
                request.Weight,
                request.AvailableForIssuance,
                0,
                DateOnly.Parse(DateTime.Now.ToShortDateString()),
                request.MinAge
            );

            return bookType;
        }
    }
}

File: easy-library-ui/src/Services/bookTypes.service.js
const API_BASE_URL = 'http://localhost:5142/api';

class BookTypesService {
    static handleResponse = async (response) => {
        if (!response.ok) {
            const error = await response.json();
            throw new Error(error.message || 'Что-то пошло не так');
        }
        return response.json();
    };

    // преобразовывает bookTypeData в формат, который ожидается сервером
    static mapToRequestFormat = (bookTypeData) => {
        bookTypeData.publishingHouseId = bookTypeData.publishingHouse;
        delete bookTypeData.publishingHouse;
        bookTypeData.bookSeriesId = bookTypeData.bookSeries;
        delete bookTypeData.bookSeries;
        bookTypeData.authorsId = bookTypeData.bookAuthors;
        delete bookTypeData.bookAuthors;
        console.log(bookTypeData);
        return bookTypeData;
    }

    static getById = async (id) => {
        const response = await fetch(`${API_BASE_URL}/booktypes/${id}`, {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json'
            },
        });
        return BookTypesService.handleResponse(response);
    }

    // type: full - все поля, display - только нужные для отображения в списке
    static getAllBookTypes = async (type) => {
        const response = await fetch(`${API_BASE_URL}/booktypes`, {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
                'Type': type
            },
        });
        return BookTypesService.handleResponse(response);
    };

    static createBookType = async (bookTypeData) => {
        const response = await fetch(`${API_BASE_URL}/booktypes`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(BookTypesService.mapToRequestFormat(bookTypeData)),
        });
        return BookTypesService.handleResponse(response);
    };

    static updateBookType = async (bookTypeData) => {
        const response = await fetch(`${API_BASE_URL}/booktypes/${bookTypeData.id}`, {
            method: 'PUT',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(BookTypesService.mapToRequestFormat(bookTypeData)),
        });
        return BookTypesService.handleResponse(response);
    };

    static deleteBookType = async (bookTypeId) => {
        const response = await fetch(`${API_BASE_URL}/booktypes/${bookTypeId}`, {
            method: 'DELETE',
            headers: {
                'Content-Type': 'application/json',
            },
        });
        return BookTypesService.handleResponse(response);
    };
}

export default BookTypesService;